language/diem-vm/mvhashmap/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::OnceCell;
+use std::{
+    cmp::{max, PartialOrd},
+    collections::{btree_map::BTreeMap, HashMap},
+    fmt::Debug,
+    hash::Hash,
+};
+
+#[cfg(test)]
+mod unit_tests;
+
+/// A structure that holds placeholders for each write to the database
+//
+//  The structure is created by one thread creating the scheduling, and
+//  at that point it is used as a &mut by that single thread.
+//
+//  Then it is passed to all threads executing as a shared reference. At
+//  this point only a single thread must write to any entry, and others
+//  can read from it. Only entries are mutated using interior mutability,
+//  but no entries can be added or deleted.
+//
+
+pub type Version = usize;
+
+pub struct MVHashMap<K, V> {
+    data: HashMap<K, BTreeMap<Version, WriteCell<V>>>,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    UnexpectedWrite,
+}
+
+#[cfg_attr(any(target_arch = "x86_64"), repr(align(128)))]
+pub(crate) struct WriteCell<V>(OnceCell<Option<V>>);
+
+impl<V: Debug> WriteCell<V> {
+    pub fn new() -> WriteCell<V> {
+        WriteCell(OnceCell::new())
+    }
+
+    pub fn is_assigned(&self) -> bool {
+        self.0.get().is_some()
+    }
+
+    pub fn write(&self, v: V) {
+        self.0
+            .set(Some(v))
+            .expect("WriteCell should be modified exactly once");
+    }
+
+    pub fn skip(&self) {
+        self.0
+            .set(None)
+            .expect("WriteCell should be modified exactly once");
+    }
+
+    pub fn get(&self) -> Option<&Option<V>> {
+        self.0.get()
+    }
+}
+
+impl<K: Hash + Clone + Eq, V: Clone + Debug> MVHashMap<K, V> {
+    /// Create the MVHashMap structure from a list of possible writes. Each element in the list
+    /// indicates a key that could potentially be modified at its given version.
+    ///
+    /// Returns the maximum number of writes that can write to one single key, and the MVHashMap.
+    pub fn new_from(possible_writes: Vec<(K, Version)>) -> (usize, MVHashMap<K, V>) {
+        let mut outer_map: HashMap<K, BTreeMap<Version, WriteCell<V>>> = HashMap::new();
+        for (key, version) in possible_writes.into_iter() {
+            outer_map
+                .entry(key)
+                .or_default()
+                .insert(version, WriteCell::new());
+        }
+        (
+            outer_map
+                .values()
+                .fold(0, |max_depth, btree_map| max(max_depth, btree_map.len())),
+            MVHashMap { data: outer_map },
+        )
+    }
+
+    /// Get the final value for each keys in the MVHashMap
+    pub fn get_change_set(&self) -> Vec<(K, V)> {
+        let mut change_set = Vec::with_capacity(self.data.len());
+        for (k, _) in self.data.iter() {
+            if let Ok(val) = self.read(k, usize::MAX) {
```

Comment:
1) if version being usize is implementation detail then maybe we also need MAX_VERSION or smt, or we could expose read_last() API that internally calls with MAX_VERSION or even usize::MAX - more contained within impl.
2) shouldn't we propagate the error or abort if the read returns an error because the entry wasn't written? (by this time everything should be skipped or written?) - Err(None) seems fine, then it was just skipped

Reply:
Now that I think about it, I think this api is actually unnecessary and even somewhat misleading. In the parallel execution, you never will need to invoke this api because you just need the output of each individual transaction. Since `K` is not ordered, the vector returned from the same insertion order may be different because of the hash randomness. Exposing this api may give people false promise that it will be a determnistic order. I'll just delete the api for now and we should return a `HashMap<K, V>` if it's needed in the future.

Reply:
good point,might be relevant in the future though if we just require the final state and not all intermediate states, but it can be added then.